apps/src/lib/node/ledger/shell/prepare_proposal/block_space_alloc/states/decrypted_txs.rs

```diff
@@ -0,0 +1,44 @@
+use std::marker::PhantomData;
+
+use super::super::{thres, AllocStatus, BlockSpaceAllocator, TxBin};
+use super::{
+    BuildingDecryptedTxBatch, BuildingProtocolTxBatch, NextStateImpl, TryAlloc,
+};
+
+impl TryAlloc for BlockSpaceAllocator<BuildingDecryptedTxBatch> {
+    #[inline]
+    fn try_alloc<'tx>(&mut self, tx: &'tx [u8]) -> AllocStatus<'tx> {
+        self.decrypted_txs.try_dump(tx)
+    }
+}
+
+impl NextStateImpl for BlockSpaceAllocator<BuildingDecryptedTxBatch> {
+    type Next = BlockSpaceAllocator<BuildingProtocolTxBatch>;
+
+    #[inline]
+    fn next_state_impl(mut self) -> Self::Next {
+        self.decrypted_txs.shrink();
+
+        // reserve space for protocol txs
+        let remaining_free_space = self.uninitialized_space_in_bytes();
+        self.protocol_txs =
```

Comment:
yeah, but then we need another 1/3 for mempool txs

Reply:
What I'm confused about is that is this 1/3 of the total block (which is what I expected), or 1/3 of the block minus the space for decrypted txs. If the latter, it would seem that decrypted gets 1/3, protocal 2/9 and mempool 4/9.

Reply:
1/3 of the total block. during the previous block, we assume we allocated about 1/3 of the space for encrypted txs, therefore decrypted txs used up 1/3 of the space. now, the remaining space should be .... fuck I see what you mean. craaap